hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/FromElementType.java

```diff
@@ -456,11 +456,15 @@ public Type getPropertyType(String propertyName, String propertyPath) {
 			// executors being used (as this subquery will
 			// actually be used in the "id select" phase
 			// of that multi-table executor)
+			// for update queries, we must only use the alias in the top level where clause
 			// B) otherwise, we need to use the persister's
 			// table name as the column qualification
 			// 2) otherwise (not correlated), use the given alias
 			if ( isCorrelation() ) {
-				if ( isMultiTable() || isInsertQuery() ) {
+				if (isMultiTable() && (!isUpdateQuery() || fromElement.getWalker().getCurrentTopLevelClauseType() == HqlSqlTokenTypes.WHERE)) {
+					return propertyMapping.toColumns(tableAlias, path);
+				}
+				else if (isInsertQuery()) {
```

Comment:
Shouldn't this still check for `isMultiTable()`?

Reply:
You mean leave it like `else if ( isMultiTable() || isInsertQuery() )`? Then you would get the same erroneous behavior as before.